crates/polars-core/src/utils/any_value.rs

```diff
@@ -0,0 +1,25 @@
+use crate::prelude::*;
+use crate::utils::try_get_supertype;
+
+/// Determine the supertype of a collection of [`AnyValue`].
+///
+/// [`AnyValue`]: crate::datatypes::AnyValue
+pub fn any_values_to_supertype<'a, I>(values: I) -> PolarsResult<DataType>
+where
+    I: IntoIterator<Item = &'a AnyValue<'a>>,
+{
+    let mut supertype = DataType::Null;
+    let mut dtypes = PlHashSet::<DataType>::new();
+    for av in values {
```

Comment:
Just for context. The indexset perf might be better, as you keep a smaller loop. Now you have a branch on every iteration of `N` and we expect `N` to be much bigger than `K`, where `N` is no. of elements and `K` is no. of unique datatypes.

Reply:
I didn't realize this type of loop would be slower - I figured the `collect` on the IndexSet would have to do something similar (check whether the dtype hash is already in the set before adding it). Any recommendations on where I can read up on this kind of stuff?

The benefit of this implementation is that it can early exit if there is no supertype (e.g. before going through all `N` values), but I guess that's relatively rare so if it makes the loop slower it's not worth it.

Reply:
The indexset is a different way of storing the values (inline in a vec, instead of the hash slots), the hash slots will store indexes to that vec. Can elaborate Thursday. ^^

> but I guess that's relatively rare so if it makes the loop slower 

I am not saying that it does :P. It is tough to tell. Though often errors in tight loops are not good. But I think the runtime is in the hashtable more than the branches here.